### ISSUE TITLE
Automatically turn on Push Notifications entitlement in iOS

### ISSUE THREAD

**Issue by slawomir-marek** [microsoft/appcenter-sdk-cordova#92]
### **Description**

It would be great if the plugin could add the aps-environment capability to the project automatically. Currently we need to enable this manually when you create an iOS platform. This is not a problem when we're building app on our dev environment. But right now we cannot build production version of our app through CI/CD which will be able to receive push notifications.

### **Repro Steps**
 1. Build app sources (ionic cordova build ios)
 2. Open project in Xcode
 3. Go to General tab 
 4. Check signing settings - Team, Provisioning Profile and Signing Certificate are ok. Push Notifications are on the included capabilities list when you display provisioning profile details
 5. Go to Capabilities tab
 6. Push Notifications are turned of

### **Details**

1. Which SDK version are you using?
0.3.4
2. Which OS version did you experience the issue on?
iOS 12.2
3. What version of Cordova, Cordova platforms, Cocoapods, and Ruby are you using?
cordova-ios: 5.0.0
4. What device version did you see this error on? Were you using a simulator or a physical device?
iPhone 5s
5. What third party plugins are you using?
cordova-android-firebase-gradle-release 3.0.0 "cordova-android-firebase-gradle-release"
cordova-android-play-services-gradle-release 3.0.0 "cordova-android-play-services-gradle-release"
cordova-plugin-actionsheet 2.3.3 "ActionSheet"
cordova-plugin-app-version 0.1.9 "AppVersion"
cordova-plugin-appcenter-analytics 0.3.4 "App Center Analytics for Cordova"
cordova-plugin-appcenter-push 0.3.4 "App Center Push for Cordova"
cordova-plugin-appcenter-shared 0.3.4 "App Center shared code for Cordova"
cordova-plugin-background-mode 0.7.2 "BackgroundMode"
cordova-plugin-browsertab 0.2.0 "cordova-plugin-browsertab"
cordova-plugin-camera 4.0.3 "Camera"
cordova-plugin-cocoapod-support 1.6.0 "Cordova CocoaPods Dependency Support"
cordova-plugin-compat 1.2.0 "Compat"
cordova-plugin-contacts 3.0.1 "Contacts"
cordova-plugin-device 2.0.2 "Device"
cordova-plugin-dialogs 2.0.1 "Notification"
cordova-plugin-inappbrowser 3.0.0 "InAppBrowser"
cordova-plugin-ionic-keyboard 2.1.3 "cordova-plugin-ionic-keyboard"
cordova-plugin-ionic-webview 3.1.2 "cordova-plugin-ionic-webview"
cordova-plugin-ms-adal 0.10.1 "ADAL for Cordova"
cordova-plugin-network-information 2.0.1 "Network Information"
cordova-plugin-splashscreen 5.0.2 "Splashscreen"
cordova-plugin-statusbar 2.4.2 "StatusBar"
cordova-plugin-whitelist 1.3.3 "Whitelist"
cordova-sqlcipher-adapter 0.2.1 "Cordova sqlcipher adapter"
uk.co.workingedge.phonegap.plugin.launchnavigator 5.0.4 "Launch Navigator"

**Comment by maciejkoch** [comment]
Hi,

I confirm this behavior and in my opinion it's a bug. It seems to be a conflict with other plugins which also are updating Entitlements-*.plist file. I use cordova-plugin-ms-adal plugin and when these plugins are installed together the push notification capability is not enabled in XCode. When I remove plugin a push notification capability is enabled. Can you check that?

**Comment by russelarms** [comment]
It was labeled feature-request by mistake. Actually, it looks like a bug.
Just to clarify, **cordova-plugin-appcenter-push** turns on push capabilities automatically.

I've tested **cordova-plugin-appcenter-push** plugin along with **cordova-plugin-ms-adal** plugin and it seems to work as expected.

I’ve used pure cordova app:

```
1. cordova create MyApp
2. cd ./MyApp && cordova platform add ios
3. cordova plugin add cordova-plugin-appcenter-push //capabilities turned on in this step.
4. cordova plugin add cordova-plugin-ms-adal //keychain-sharing capability was added.
```
I did the same with ionic cordova:
```
1. ionic start helloWorld tabs --type=ionic-angular
2. cd ./helloWorld && ionic cordova build ios
3. cordova plugin add cordova-plugin-appcenter-push  //capabilities turned on in this step.
4. cordova plugin add cordova-plugin-ms-adal //keychain-sharing capability was added.

```
Could you please perform ‘pod repo update’ and verify if it works with the latest plugin version (0.3.5)?
Also could you please verify if it works for you with the steps I wrote above?

**Comment by maciejkoch** [comment]
Hi and thanks for your reply.

I checked that with your steps and it doesn't seem to be working as expected. After adding cordova-plugin-appcenter-push notification capability was turned on but when I added cordova-plugin-ms-adal notification capability was turned off again.

BTW, I use pipeline in AzureDevops to build the app so all plugins are installed based on package.json and config.xml. It means I'm not able to install plugins step by step.

**Comment by russelarms** [comment]
@maciejkoch Does this problem occur only with the ms-adal plugin, or are there other plugins that may cause this problem?
I mean, it looks like the issue may be in the ms-adal plugin itself.

**Comment by maciejkoch** [comment]
@russelarms I didn't find other plugins that may cause the problem. I think the issue is related to a fact that both plugins modify "*-Debug.plist" and "*-Release.plist" files. If ms-adal plugin is no longer maintained is there any possibility to take a look on that problem? I'm asking because both of them are created by MS.

**Comment by russelarms** [comment]
I'm investigating how different plugins may influent each other.

@maciejkoch BTW could you substitute ms-adal plugin with this library: https://github.com/AzureAD/microsoft-authentication-library-for-js
Does it meet your requirements?

**Comment by russelarms** [comment]
@maciejkoch I've tested how two different cordova plugins modify ios entitlements and haven't found any issues. I've applied **appcenter-push** plugin and local version of **ms-adal** `plugins.xml` to the same demo application.

**Comment by russelarms** [comment]
As there is no response in a while I'm closing this issue. Please reopen/comment if you have more questions or run into this again.

**Comment by slawomir-marek** [comment]
Is there any chance to share this demo app with us? Maybe this bug is related with iOS or XCode version? This is really strange...

**Comment by russelarms** [comment]
@slawomir-marek I'm afraid I no longer have this sample. Can you verify step-by-step how the capabilities behave when doing the steps I mentioned above:

`1. ionic start helloWorld tabs --type=ionic-angular`
`2. cd ./helloWorld && ionic cordova build ios`
`3. cordova plugin add cordova-plugin-appcenter-push  //capabilities turned on in this step.`
`4. cordova plugin add cordova-plugin-ms-adal //keychain-sharing capability was added.`

**Comment by slawomir-marek** [comment]
@russelarms I'll check it tomorrow and I'll let you know.

**Comment by russelarms** [comment]
@slawomir-marek do you have any updates?

**Comment by slawomir-marek** [comment]
Hi @russelarms . Again sorry for delay.

I can confirm that I built this sample project once with success - push notification capability was tuned on. But when I removed platform folder and tried to build project again capability was turned off. I need more time to check it. 

I had also problem with adding _cordova-plugin-appcenter-push_ into project. I received error
``` pod command failed with exit code 1 ```
but I found explanation [here](https://docs.microsoft.com/en-us/appcenter/sdk/troubleshooting/cordova#failed-to-install-plugin---pod-command-failed-with-exit-code-1)

**Comment by slawomir-marek** [comment]
Hi again @russelarms

I made some additional tests. I modified your example a little bit to be sure that demo app will be more similar to my project. First of all I created app based on Ionic 4, so without this ```--type=ionic-angular``` parameter. Secondly I used cordova-ios in version 5.0.0. 

Steps to reproduce:
```
ionic start helloWorld tabs
cd ./helloWorld
ionic cordova build ios 

cordova platform add ios@5.0.0
rm -rf platforms && rm -rf plugins
ionic cordova build ios

cordova plugin add cordova-plugin-appcenter-push
// I checked capabilities tab in xCode - push notifications was tuned on

rm -rf platforms && rm -rf plugins
ionic cordova build ios
// I checked capabilities tab in xCode - push notifications was tuned on

cordova plugin add cordova-plugin-ms-adal
// I checked capabilities tab in xCode - push notifications was tuned off

cordova plugin remove cordova-plugin-ms-adal
// I checked capabilities tab in xCode - push notifications was tuned on
```

**Comment by russelarms** [comment]
@slawomir-marek I've tried the steps you wrote but wasn't able to reproduce the issue. Capabilities just behaved as expected. Could you share your environment, like the versions of:
-`node --version`
-`npm --version`
-`cordova --version`
-`ionic --version` //4.12.0 I guess?
-Xcode and Xcode CLT versions -  `/Applications/Xcode.app/Contents/Developer/usr/bin/xcodebuild -version`

**Comment by slawomir-marek** [comment]
@russelarms:
 - node: 8.10.0
 - npm: 5.6.0
 - cordova: 8.1.2
 - ionic: 5.2.3 (but this is version of ionic-cli not ionic framework)
 - Xcode: 10.2.1 build version 10E1001

**Comment by annakocheshkova** [comment]
Hey @slawomir-marek, I just checked this on your environment. It reproduced for me, although the combination of node 8 and ionic 5.2.3 didn't work so I upgraded to 10. Nevermind.
I looked at the `ms-adal` post scripts and found [this one](https://github.com/AzureAD/azure-activedirectory-library-for-cordova/blob/archive/scripts/configureEntitlementsIos.js). This script is adding the Entitlements file which overwrites the Push notifications capability. The repository is archived so I can't even make a PR to fix this somehow. 

The solution would be to either install push plugin after the `ms-adal` plugin, or turn them on manually (open ios project in *XCode*, go to *Capabilities* tab > turn on Push notifications).
